### DISCUSSION TITLE
A suite of add-on components for Jamulus server (gojam, jamulus-lounge, jamviz)

### DISCUSSION THREAD

**Discussion by dtinth** [jamulussoftware/jamulus#3085] [Ideas]
I’m running a Jamulus server in Thailand ([MusicJammingTH.net](https://musicjammingth.net)) and I have built a few add-on components for it. I’m sharing them here in case other server owners find them useful.

> **Note**
> Currently, these components are built for Linux servers only, and requires Docker and programming knowledge to set up on your own server.

<br>

## Overview of components

<img width="1318" alt="image" src="https://github.com/jamulussoftware/jamulus/assets/193136/608aaa36-4270-4dc3-8831-5627ca2d625f">

- [jamviz](https://github.com/dtinth/jamviz) is a Jamulus activity visualizer, useful for creating nice visuals for a livestream / recording. ([See example video.](https://youtu.be/0-ZXr-3cpcI))
  - Requires _jamulus-lounge_ to be setup on the server.
- [jamulus-lounge](https://github.com/dtinth/jamulus-lounge) provides a web-based interface to a Jamulus server to let people listen in from a web browser.
  - Built on top of _gojam_.
- [gojam](https://github.com/dtinth/gojam) is an implementation of a Jamulus client written in Go language.

<br>

## gojam

**gojam** is a programmable Jamulus client written in Go language. Once connected to a server, it can generate a live MP3 audio stream and JSON event stream. It is used as a foundation for the next component, _jamulus-lounge_. It has a very limited feature set (just enough to support _jamulus-lounge_), but can be used standalone to build bots and other kind of integrations. Source code here: [dtinth/gojam](https://github.com/dtinth/gojam).

<br>

## jamulus-lounge

**jamulus-lounge** is a web-based interface for `gojam`. You can install jamulus-lounge on your own Jamulus server to let people listen in from a web browser. Source code here: [dtinth/jamulus-lounge](https://github.com/dtinth/jamulus-lounge).

<img width="1221" alt="image" src="https://github.com/jamulussoftware/jamulus/assets/193136/3bcb5a90-1e1e-437f-b1ba-40be5ac4386b">

&nbsp;

## jamviz

**jamviz** is a Jamulus activity visualizer. It takes the data from jamulus-lounge, and generates a live (or replayed) visualization of Jamulus activity. It looks like this:

<img width="1982" alt="image" src="https://github.com/jamulussoftware/jamulus/assets/193136/f082b1cc-8149-45e1-8c24-d1056394ee27">

It can be used to create nice visuals for a livestream (or a recording) of a Jamulus session. [See example video.](https://youtu.be/0-ZXr-3cpcI)

It is a web-based application hosted at <https://jamviz.vercel.app/> and the source code is here: [dtinth/jamviz](https://github.com/dtinth/jamviz)

**Comment by ann0see** [comment]
Interesting and thanks for sharing! Is there a reason why you re-wrote the client instead of modifying the existing C++ client? Probably you could add more customization to your client?
I seem to understand that @mcfnord is interested in your work.

**Reply by dtinth** [reply]
A few reasons:

- I'm more comfortable writing Go than C++. I find using goroutines easier to work with than threads or signals. And I'm not very productive in languages without a garbage collector.
- gojam is built for headless use case (no UI / direct sound output) and intended to be used programmatically. So I avoid big dependencies like audio libraries, Qt, etc.

**Reply by ann0see** [reply]
Thanks for the comment. 

Hoffie once proposed to have a simplified "recorder" or API client - which is exactly what you did - at least to some extent. 
I'm just thinking out loudly, but I think of gojam a good foundation for further work. 

The website based approach has the upside that once setup everyone can use it without much hassle. However, some people might want to have a simple local client for listening and recording for arbitrary servers - and want to have it on their own machine. I believe gojam could be used as backend.

**Comment by ann0see** [comment]
For all - not just this, feel free to update https://jamulus.io/kb/2022/01/25/Related-Projects.html

**Reply by mcfnord** [reply]
https://github.com/jamulussoftware/jamulus/pull/3090

**Comment by mcfnord** [comment]
I added a leasing mechanism to one instance of jamulus-lounge to let it connect to most servers. This lets me notice Ultimate Guitar links and display the song title/artist. Perhaps eventually I can do same with SSE streams for the stationary lounges (3 right now). I'd like to support Jitsi links in a similar way. I hope to detect silence by watching the level data. I've built and run gojam on Raspberry Pi 4 to listen in. (Not sure what the input device could be.)
